File: api/src/main/java/me/kvalbrus/multibans/api/managers/PunishmentManager.java
package me.kvalbrus.multibans.api.managers;

import java.util.List;
import java.util.UUID;
import me.kvalbrus.multibans.api.Player;
import me.kvalbrus.multibans.api.punishment.Punishment;
import me.kvalbrus.multibans.api.punishment.PunishmentType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface PunishmentManager {

    @NotNull
    <T extends Punishment> List<T> getPlayerHistory(UUID uuid);

    @NotNull
    <T extends Punishment> List<T> getPlayerHistory(UUID uuid, Class<T> clazz);

    @NotNull
    <T extends Punishment> List<T> getCreatorHistory(String creator);

    @NotNull
    <T extends Punishment> List<T> getActivePunishments(UUID uuid);

    @NotNull
    <T extends Punishment> List<T> getActivePunishments(UUID uuid, Class<T> clazz);

    boolean hasPunishment(String id);

    @Nullable
    Punishment getPunishment(String id);

    boolean hasActiveBan(UUID uuid);

    boolean hasActiveBan(String name);

    boolean hasActiveBanIp(UUID uuid);

    boolean hasActiveBanIp(String name);

    boolean hasActiveChatMute(UUID uuid);

    boolean hasActiveChatMute(String name);

    @NotNull
    Punishment generatePunishment(@NotNull final PunishmentType type,
                                  @NotNull final Player target,
                                  @NotNull final String creator,
                                  long duration,
                                  @NotNull final String reason);

    @NotNull
    Punishment generatePunishment(@NotNull final PunishmentType type,
                                  @NotNull final Player target,
                                  @NotNull final String creator,
                                  long duration,
                                  @NotNull final String reason,
                                  @Nullable String comment,
                                  @NotNull List<String> servers);
}

File: api/src/main/kotlin/me/kvalbrus/multibans/api/managers/PunishmentManager.kt
package me.kvalbrus.multibans.api.managers

import me.kvalbrus.multibans.api.punishment.Punishment
import me.kvalbrus.multibans.api.punishment.executor.PunishmentExecutor
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentType
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget
import java.util.*
import kotlin.jvm.Throws

interface PunishmentManager {

    @Throws(Exception::class)
    fun <T : Punishment> getPlayerHistory(name: String) : List<T>

    @Throws(Exception::class)
    fun <T : Punishment> getPlayerHistory(uuid: UUID) : List<T>

    @Throws(Exception::class)
    fun <T : Punishment> getPlayerHistory(uuid: UUID, clazz: Class<T>): List<T>

    @Throws(Exception::class)
    fun <T : Punishment> getCreatorHistory(creatorName: String): List<T>

    @Throws(Exception::class)
    fun <T : Punishment> getActivePunishments(uuid: UUID): List<T>

    @Throws(Exception::class)
    fun <T : Punishment> getActivePunishments(uuid: UUID, clazz: Class<T>): List<T>

    @Throws(Exception::class)
    fun hasPunishment(id: String): Boolean

    @Throws(Exception::class)
    fun getPunishment(i: String): Punishment

    @Throws(Exception::class)
    fun hasActiveBan(uuid: UUID): Boolean

    @Throws(Exception::class)
    fun hasActiveBan(name: String): Boolean

    @Throws(Exception::class)
    fun hasActiveBanIp(uuid: UUID): Boolean

    @Throws(Exception::class)
    fun hasActiveBanIp(name: String): Boolean

    @Throws(Exception::class)
    fun hasActiveChatMute(uuid: UUID): Boolean

    @Throws(Exception::class)
    fun hasActiveChatMute(name: String): Boolean

    @Throws(Exception::class)
    fun generatePunishment(
        type: PunishmentType,
        target: PunishmentTarget,
        creator: PunishmentExecutor,
        duration: Long,
        reason: String
    ): Punishment

    @Throws(Exception::class)
    fun generatePunishment(
        type: PunishmentType,
        target: PunishmentTarget,
        creator: PunishmentExecutor,
        duration: Long,
        reason: String,
        comment: String?,
        servers: List<String?>
    ): Punishment
}